File: server/chat-orchestrator.js
/* =========================================
    チャット処理オーケストレーター
   ========================================= */

/*
## 概要
チャット処理の全体フローを管理し、各専門サービスを協調させる責任を持つ。
server.jsから呼ばれるメインエントリーポイント。

## 責任
- チャット処理フローの管理
- サービス間の協調
- エラー時のフォールバック制御
- レスポンス統合
*/

import { AgentDispatcher } from './agent-dispatcher.js';
import { ConversationManager } from './conversation-manager.js';
import { ContextBuilder } from './context-builder.js';
import { CommandValidator } from './command-validator.js';

export class ChatOrchestrator {
    constructor() {
        this.agentDispatcher = new AgentDispatcher();
        this.conversationManager = new ConversationManager();
        this.contextBuilder = new ContextBuilder();
        this.commandValidator = new CommandValidator();
    }

    /**
     * メインのチャット処理エントリーポイント
     */
    async processChat(message, provider = 'claude', model, context = {}) {
        try {
            // 1. 入力バリデーション
            this._validateInput(message, provider, model, context);

            // 2. 会話コンテキストの準備
            const enrichedContext = await this.conversationManager.prepareContext(context);
            
            // 3. 新しいチャット提案の判定
            const shouldSuggestNewChat = this.conversationManager.shouldSuggestNewChat(enrichedContext);

            // 4. エージェント処理の実行
            const agentResult = await this.agentDispatcher.dispatch(
                message, 
                provider, 
                model, 
                enrichedContext
            );

            // 5. コマンドの検証
            const validatedCommands = await this._validateCommands(agentResult.commands);

            // 6. 最終レスポンスの構築
            return this.contextBuilder.buildSuccessResponse({
                ...agentResult,
                commands: validatedCommands,
                shouldSuggestNewChat,
                historyCount: enrichedContext.conversationHistory?.length || 0,
                customPromptUsed: !!(enrichedContext.customPrompt?.enabled)
            });

        } catch (error) {
            console.error('❌ Chat Orchestrator Error:', error);
            
            // エラー時のフォールバック処理
            return this.contextBuilder.buildErrorResponse(
                error, 
                provider, 
                model, 
                message, 
                context
            );
        }
    }

    /**
     * 入力パラメータの基本バリデーション
     */
    _validateInput(message, provider, model, context) {
        if (!message || typeof message !== 'string') {
            throw new Error('メッセージが無効です');
        }

        if (provider && typeof provider !== 'string') {
            throw new Error('プロバイダーは文字列である必要があります');
        }

        if (model && typeof model !== 'string') {
            throw new Error('モデルは文字列である必要があります');
        }

        if (context && typeof context !== 'object') {
            throw new Error('コンテキストはオブジェクトである必要があります');
        }
    }

    /**
     * コマンド配列の検証
     */
    async _validateCommands(commands) {
        if (!Array.isArray(commands)) {
            return [];
        }

        const validatedCommands = [];
        for (const command of commands) {
            try {
                if (this.commandValidator.validate(command)) {
                    validatedCommands.push(command);
                }
            } catch (validationError) {
                console.warn('❌ Command validation failed:', validationError.message);
                // 無効なコマンドはスキップして処理を継続
            }
        }

        return validatedCommands;
    }

    /**
     * 利用可能なエージェント一覧を取得
     */
    getAvailableAgents() {
        return this.agentDispatcher.getAvailableAgents();
    }

    /**
     * システム状態の取得
     */
    getSystemStatus() {
        return {
            agents: this.agentDispatcher.getAvailableAgents(),
            conversationManager: this.conversationManager.getStatus(),
            timestamp: new Date().toISOString()
        };
    }

    /**
     * 検索履歴を取得
     */
    getSearchHistory(limit = 10) {
        return this.agentDispatcher.getSearchHistory(limit);
    }

    /**
     * 検索履歴をクリア
     */
    clearSearchHistory() {
        return this.agentDispatcher.clearSearchHistory();
    }

    /**
     * 検索サービスの状態を取得
     */
    getSearchServiceStatus() {
        return this.agentDispatcher.getSearchServiceStatus();
    }
}

File: server_pythonver/chat/chat_orchestrator.py
import datetime
import json
from ..ai.agent_dispatcher import AgentDispatcher
from ..chat.conversation_manager import ConversationManager
from ..validation.command_validator import CommandValidator
from ..validation.response_builder import ResponseBuilder # ResponseBuilderをインポート
from ..tool.web_search_service import WebSearchService # WebSearchServiceをインポート
from ..utils.response_utils import get_mock_response # 必要に応じて利用

class ChatOrchestrator:
    def __init__(self, agent_dispatcher: AgentDispatcher, conversation_manager: ConversationManager, command_validator: CommandValidator, response_builder: ResponseBuilder, web_search_service: WebSearchService):
        self.agent_dispatcher = agent_dispatcher
        self.conversation_manager = conversation_manager
        self.command_validator = command_validator
        self.response_builder = response_builder
        self.web_search_service = web_search_service # WebSearchServiceを追加

    async def process_chat(self, message: str, provider: str = "gemini", model: str = None, context: dict = None):
        if context is None:
            context = {}
            
        try:
            # 1. 入力バリデーション
            if not message or not isinstance(message, str):
                raise ValueError("メッセージが無効です")

            # 2. 会話コンテキストの準備
            enriched_context = await self.conversation_manager.prepare_context(context)
            if enriched_context is None:
                enriched_context = {}

            # 3. エージェント処理の実行
            agent_result = await self.agent_dispatcher.dispatch(
                message,
                enriched_context,
                provider,
                model
            )
            if agent_result is None:
                return self.response_builder.build_fallback_response(
                    message,
                    enriched_context,
                    reason="Agent dispatcher failed to return a result."
                )

            # 4. コマンドの検証と実行
            validated_commands = self._validate_commands(agent_result.get("commands", []))
            executed_command_results = []
            search_results_data = None

            if validated_commands:
                for command in validated_commands:
                    action = command.get("action")
                    if action == "web_search":
                        query = command.get("query")
                        options = command.get("options", {})
                        if query:
                            search_result = await self.web_search_service.perform_search(query, options)
                            executed_command_results.append(search_result)
                            if search_result.get("success"):
                                search_results_data = search_result.get("results")
                    else:
                        # 将来的に他のコマンドを扱う場合
                        executed_command_results.append({"command": command, "status": "skipped", "message": "Action not executable."})

            # フロントエンドに実行させるコマンドのみをフィルタリング
            frontend_commands = [
                cmd for cmd in validated_commands 
                if cmd.get("action") != "web_search"
            ]

            # 5. 最終レスポンスの構築
            custom_prompt = enriched_context.get("customPrompt") or {}
            response_data = {
                "message": agent_result.get("message", "処理が完了しました。"),
                "commands": frontend_commands,
                "executed_command_results": executed_command_results,
                "search_results": search_results_data,
                "agent_used": agent_result.get("agent"),
                "raw_llm_response": agent_result.get("raw_llm_response"),
                "provider": provider,
                "model": model,
                "parse_success": agent_result.get("parse_success", True),
                "warning": agent_result.get("warning"),
                "error": agent_result.get("error"),
                "should_suggest_new_chat": self.conversation_manager.should_suggest_new_chat(enriched_context),
                "custom_prompt_used": bool(custom_prompt.get("enabled")),
                "custom_prompt_name": custom_prompt.get("name"),
                "debug_context": enriched_context
            }
            return self.response_builder.build_success_response(response_data)

        except Exception as e:
            print(f"❌ Chat Orchestrator Error: {e}")
            safe_context = context if context is not None else {}
            return self.response_builder.build_error_response(
                e,
                provider,
                model,
                message,
                safe_context
            )

    def _validate_commands(self, commands: list) -> list:
        if not isinstance(commands, list):
            return []

        validated_commands = []
        for cmd in commands:
            try:
                if self.command_validator.validate(cmd):
                    validated_commands.append(cmd)
            except ValueError as e:
                print(f"Command validation failed: {e} for command: {cmd}")
        return validated_commands

    def get_available_agents(self):
        return self.agent_dispatcher.get_available_agents()

    def get_system_status(self):
        # import datetime # 不要（トップレベルでインポート済み）
        return {
            "agent_dispatcher": self.agent_dispatcher.get_status(),
            "conversation_manager": self.conversation_manager.get_status(),
            "command_validator": self.command_validator.get_validator_stats(),
            "response_builder": self.response_builder.get_response_stats(), # ResponseBuilderのステータスを追加
            "timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat()
        }